openspec/changes/add-model-provider-cred-store-oagw/specs/model-provider/spec.md

```diff
@@ -0,0 +1,633 @@
+
+### Requirement: model_provider client caching
+The system SHALL cache resolved provider client instances for performance.
+
+Implementation SHOULD use in-memory cache with:
+- **Max capacity:** 1,000 client instances
+- **TTL:** 300 seconds (5 minutes)
+- **Cache key:** `(tenant_id, provider_id, plugin_id)`
+
+#### Scenario: Client cache hit
+- **WHEN** client resolution requested for cached provider
+- **THEN** returns cached client instance (no plugin resolution)
+- **AND** records cache hit metric
+
+#### Scenario: Client cache eviction
+- **WHEN** provider configuration is updated
+- **THEN** invalidate cached client for that provider
+- **AND** next resolution creates fresh client
+
+### Requirement: model_provider OData support
+The system SHALL implement OData query capabilities per ModKit conventions.
+
+REST endpoints SHALL support:
+- `$skip` and `$top` for pagination
+- `$filter` for filtering
+- `$orderby` for sorting
+- `$select` for field projection
+
+Implementation SHALL use `modkit::api::OperationBuilder` OData support.
+
+#### Scenario: Paginated provider list
+- **WHEN** `GET /model-provider/v1/providers?$skip=20&$top=10`
+- **THEN** returns providers 21-30
+- **AND** includes `@odata.nextLink` for pagination
+
+#### Scenario: Filter by vendor
+- **WHEN** `GET /model-provider/v1/providers?$filter=vendor_gts_id eq 'gts.x.genai.provider.vendor.v1~openai'`
+- **THEN** returns only OpenAI providers
+
+#### Scenario: Filter inherited providers
+- **WHEN** `GET /model-provider/v1/providers?$filter=available_to_children eq true`
+- **THEN** returns providers shared by parent tenants
+
+### Requirement: model_provider timeout configuration
+The system SHALL support configurable timeouts for provider invocations.
+
+#### Extended table: genai_model_provider_endpoint
+- Add: `connection_timeout_ms int not null default 5000` (5s to establish connection)
+- Add: `request_timeout_ms int not null default 30000` (30s total request duration)
+- Add: `idle_timeout_ms int not null default 60000` (60s idle connection reuse)
+
+These settings are overriable by OAGW per-tenant, per-endpoint and per-link defaults.
+
+#### Scenario: Connection timeout
+- **WHEN** connection to provider takes > 5s (configured in endpoint)
+- **THEN** gateway aborts connection attempt
+- **AND** returns 504 Gateway Timeout to client
+
+#### Scenario: Request timeout
+- **WHEN** model invocation takes > 30s (configured in endpoint)
+- **THEN** gateway cancels request
+- **AND** returns 504 Gateway Timeout to client
+
+### Requirement: model_provider plugin trait definition
+The system SHALL define a Rust trait for provider plugins.
+
+```rust
+#[async_trait]
+pub trait ModelProviderPlugin: Send + Sync {
+    /// Plugin metadata
+    fn metadata(&self) -> &PluginMetadata;
+
+    /// Check if plugin can handle this provider configuration
+    fn can_handle(&self, vendor_gts_id: &str, kind_gts_id: &str, endpoint_type_gts_id: &str) -> bool;
```

Comment:
Looks like this call doesn't support all model selection criteries:

- capabilities
- features
- tenant requirements
- quotas
- priority